File: C++/first_unique_character_in_a_string.cpp
// no. 387
#include <map>
#include <string>
#include <vector>
using namespace std;

class Solution {
public:
  int firstUniqChar(string s) {
    map<char, vector<int>> counter;
    for (int i = 0; i < s.length(); i++) {
      counter[s[i]].push_back(i);
    }
    int first_index = s.length();
    map<char, vector<int>>::iterator it;
    for (it = counter.begin(); it != counter.end(); it++) {
      if ((it->second.size() == 1) & (it->second[0] < first_index)) {
        first_index = it->second[0];
      }
    }
    return first_index == s.length() ? -1 : first_index;
  }
};

File: python/first_unique_character_in_a_string.py
#no. 387
class Solution:
    def firstUniqChar(self, s: str) -> int:
        counter = {c: [] for c in 'abcdefghijklmnopqrstuvwxyz'}
        for i in range(len(s)):
            counter[s[i]].append(i)
        first_index = len(s)
        for c in counter:
            if len(counter[c]) == 1:
                if counter[c][0] < first_index:
                    first_index = counter[c][0]
        return -1 if first_index == len(s) else first_index
